File: lang/nodejs/src/fun/vac/data_structures/CircularlyLinkedList.js
"use strict"

let Node = class {

    constructor(element) {
        this.data = element
        this.next = null
    }
}

exports.CircularlyLinkedList = class {

    constructor(physicalSize = 0) {
        this.tail = null
        this.length = 0
    }

    size() {
        return this.length
    }

    isEmpty() {
        return this.length === 0
    }

    get(index) {
        if (index < 0 || index >= this.length) {
            throw new Error("[PANIC - IndexOutOfBounds]")
        }

        let cursor = this.tail

        if (index !== this.length - 1) {
            for (let i = 0; i <= index; i++) {
                cursor = cursor.next
            }
        }

        return cursor.data
    }

    set(index, element) {
        if (index < 0 || index >= this.length) {
            throw new Error("[PANIC - IndexOutOfBounds]")
        }

        let cursor = this.tail

        if (index !== this.length - 1) {
            for (let i = 0; i <= index; i++) {
                cursor = cursor.next
            }
        }

        cursor.data = element
    }

    insert(index, element) {
        if (index < 0 || index > this.length) {
            throw new Error("[PANIC - IndexOutOfBounds]")
        }

        let node = new Node(element)

        if (index === 0) {
            if (this.length === 0) {
                node.next = node
                this.tail = node
            } else {
                node.next = this.tail.next
                this.tail.next = node
            }
        } else if (index === this.length) {
            node.next = this.tail.next
            this.tail.next = node
            this.tail = node
        } else {
            let cursor = this.tail
            for (let i = 0, bound = index - 1; i <= bound; i++) {
                cursor = cursor.next
            }
            node.next = cursor.next
            cursor.next = node
        }

        this.length += 1
    }

    remove(index) {
        if (index < 0 || index >= this.length) {
            throw new Error("[PANIC - IndexOutOfBounds]")
        }

        let target = null

        if (index === 0) {
            target = this.tail.next
            if (this.length === 1) {
                this.tail = null
            } else {
                this.tail.next = target.next
            }
        } else {
            let cursor = this.tail
            for (let i = 0, bound = index - 1; i <= bound; i++) {
                cursor = cursor.next
            }
            target = cursor.next
            cursor.next = target.next
            if (index === this.length - 1) {
                this.tail = cursor
            }
        }

        target.data = null

        this.length -= 1
    }

    front() {
        return this.get(0)
    }

    back() {
        return this.get(this.length - 1)
    }

    prepend(element) {
        this.insert(0, element)
    }

    append(element) {
        this.insert(this.length, element)
    }

    poll() {
        this.remove(0)
    }

    eject() {
        this.remove(this.length - 1)
    }
}

File: go/muse/data_structures/circularly_linked_list/circularly_linked_list.go
package circularly_linked_list

import "errors"

type Node[T any] struct {
	data T
	next *Node[T]
}

type CircularlyLinkedList[T any] struct {
	tail   *Node[T]
	length int
}

func New[T any]() *CircularlyLinkedList[T] {
	return &CircularlyLinkedList[T]{tail: nil, length: 0}
}

func (l *CircularlyLinkedList[T]) Size() int {
	return l.length
}

func (l *CircularlyLinkedList[T]) IsEmpty() bool {
	return l.length == 0
}

func (l *CircularlyLinkedList[T]) Get(index int) T {
	if index < 0 || index >= l.length {
		panic(errors.New("[PANIC - IndexOutOfBounds]"))
	}

	cursor := l.tail

	if index != l.length-1 {
		for range index + 1 {
			cursor = cursor.next
		}
	}

	return cursor.data
}

func (l *CircularlyLinkedList[T]) Set(index int, element T) {
	if index < 0 || index >= l.length {
		panic(errors.New("[PANIC - IndexOutOfBounds]"))
	}

	cursor := l.tail

	if index != l.length-1 {
		for range index + 1 {
			cursor = cursor.next
		}
	}

	cursor.data = element
}

func (l *CircularlyLinkedList[T]) Insert(index int, element T) {
	if index < 0 || index > l.length {
		panic(errors.New("[PANIC - IndexOutOfBounds]"))
	}

	node := &Node[T]{data: element, next: nil}

	switch index {
	case 0:
		if l.length == 0 {
			node.next = node
			l.tail = node
		} else {
			node.next = l.tail.next
			l.tail.next = node
		}
	case l.length:
		node.next = l.tail.next
		l.tail.next = node
		l.tail = node
	default:
		cursor := l.tail
		for range index {
			cursor = cursor.next
		}
		node.next = cursor.next
		cursor.next = node
	}

	l.length++
}

func (l *CircularlyLinkedList[T]) Remove(index int) {
	if index < 0 || index >= l.length {
		panic(errors.New("[PANIC - IndexOutOfBounds]"))
	}

	var target *Node[T]

	if index == 0 {
		target = l.tail.next
		if l.length == 1 {
			l.tail = nil
		} else {
			l.tail.next = target.next
		}
	} else {
		cursor := l.tail
		for range index {
			cursor = cursor.next
		}
		target = cursor.next
		cursor.next = target.next
		if index == l.length-1 {
			l.tail = cursor
		}
	}

	target.data = *new(T)

	l.length--
}

func (l *CircularlyLinkedList[T]) Front() T {
	return l.Get(0)
}

func (l *CircularlyLinkedList[T]) Back() T {
	return l.Get(l.length - 1)
}

func (l *CircularlyLinkedList[T]) Prepend(element T) {
	l.Insert(0, element)
}

func (l *CircularlyLinkedList[T]) Append(element T) {
	l.Insert(l.length, element)
}

func (l *CircularlyLinkedList[T]) Poll() {
	l.Remove(0)
}

func (l *CircularlyLinkedList[T]) Eject() {
	l.Remove(l.length - 1)
}
